README.md
# Timetable.js
Vanilla javascript plugin for building nice responsive timetables. Provides a simple javascript interface to add events and locations which can be rendered to nice HTML. Works on mobile devices as well.

## Just show me the demo
Okay: [**demo**](http://timetablejs.grible.co). Also, have a look at the website: [http://timetablejs.org](http://timetablejs.org).

## Installation
Install with bower, or alternatively download the [ZIP](https://github.com/Grible/timetable.js/archive/master.zip):

```
bower install timetable
```

Load the plugin and styles in your HTML from the `dist` folder:
```html
<link rel="stylesheet" href="timetablejs.css">

<script src="timetable.min.js"></script>
```
Add a timetable placeholder:
```html
<div class="timetable"></div>
```

## Usage
Make a timetable object, optionally set the scope in hours (the visible hours in the timetable):
```javascript
var timetable = new Timetable();
timetable.setScope(9, 17); // optional, only whole hours between 0 and 23
```
Add some locations:
```javascript
timetable.addLocations(['Silent Disco', 'Nile', 'Len Room', 'Maas Room']);
```
Add your events using `addEvent(name, location, startDate, endDate[, options])`:
```javascript
timetable.addEvent('Frankadelic', 'Nile', new Date(2015,7,17,10,45), new Date(2015,7,17,12,30));
```

In addition, you can pass options through an object (optional):
```javascript
var options = {
  url: '#', // makes the event clickable
  class: 'vip', // additional css class
  data: { // each property will be added to the data-* attributes of the DOM node for this event
    id: 4,
    ticketType: 'VIP'
  },
  onClick: function(event, timetable, clickEvent) {} // custom click handler, which is passed the event object and full timetable as context  
};
timetable.addEvent('Jam Session', 'Nile', new Date(2015,7,17,21,15), new Date(2015,7,17,23,30), options);
 ```

Last, render the thing in your previously created timetable placeholder:
```javascript
var renderer = new Timetable.Renderer(timetable);
renderer.draw('.timetable'); // any css selector
```
That's it!

## Changing the looks
Instead of adding the `timetablejs.css` directly to your HTML, you could import `app/styles/plugin.sass` to your own SASS file. All colors and spacing values are defined as default variables which you can easily override. These are the defaults:
```sass
// general
$timetable-use-sticky-header: false !default

// dimensions
$timetable-hour-column-width: 96px !default
$timetable-hour-row-height: 46px !default
$timetable-heading-height: 30px !default

// colors & decoration
$timetable-grid-color: #E5E5E5 !default
$timetable-grid: 1px solid $timetable-grid-color !default
$timetable-row-header-padding: 15px !default
$timetable-row-header-color: #EFEFEF !default
$timetable-legend-row-separator: 1px solid white !default
$timetable-entry-row-separator: none !default
$timetable-row-header-gap: 5px solid transparent !default
$timetable-row-uneven-color: #FDFDFD !default
$timetable-row-even-color: #F4F4F4 !default
$timetable-entry-color: #EC6A5E !default
$timetable-entry-color-hover: darken($timetable-entry-color, 10%) !default
$timetable-entry-border: 1px solid darken($timetable-entry-color, 15%) !default
$timetable-entry-padding: 10px !default
$timetable-background-color: white !default
```

The option `$timetable-use-sticky-header` makes the time indicator stick to the top of the screen while scrolling down, which can be very pleasant with large timetables or on mobile devices. Note that this is not yet widely supported in browsers, although all modern (non-IE) browsers support it. Browsers that don't support it will fall back to a non-sticky header automatically. [More info here](https://caniuse.com/#feat=css-sticky). 

Alternatively you could override the css styles manually.

## Browser support
Timetable.js has been designed to work with modern browsers (only). It has been tested with the latest version of the most common browsers.

## Contributing
Please use the Github issue tracker for issues/feature requests. We use Gulp for development and Mocha with Chai for unit testing. The styles are defined in SASS. Feel free to comment/contribute.


CHANGELOG
0.8.0
- added support for sticky header: time indicator scrolling along when the timetable is larger than the viewport
- converted remaining floating styles to flexbox

0.7.0
- added support for a click handler, called when an event in the timetable is clicked
- small styling update based on flexbox
- cleanup & update of build system

0.6.0
- added support for adding event options such as URLs, classes, and data attributes
- a few bug fixes

0.5.0
- possibility to add a row seperation border with sass
- readme updates
- some bugfixes

0.4.0
- made event URLs optional to support non-clickable events

0.3.0
- added support for timetable scopes that span 2 days (e.g. from 23.00 to 4.00), for people that like night events. Maximum span is still 24 hours (e.g. from 8.00 to 8.00)

0.2.0
- removed dependency on bourbon, thereby now free of sass/css dependencies, easier for including sass in other projects

0.1.0
- first operative version


test/spec/temp.js
(function () {
  'use strict';

  describe('The test tester', function () {

    it('should run here few assertions', function () {
      expect(13).to.be.equal(10+3);
    });
  });
})();


test/spec/test.js
(function () {
  'use strict';

  beforeEach(function() {
    this.timetable = new Timetable();
  });

  describe('Timetable', function () {

    it('should by default set its scope from 9 to 17 hours', function () {
      expect(this.timetable.scope).to.deep.equal({hourStart: 9, hourEnd: 17});
    });

    it('should correct work with different dates', function(){
      var errorMsg = 'Wrong date format. Use JavaScript date format.';
      var errorFormat = 'Wrong format. Use JavaScript date format';
      var tt = this.timetable;
      expect(function(){tt.setDate('sdfhjskdhf'); }).to.throw(Error, errorMsg);
      //expect(function(){tt.setDate('28 11 2016', 'dsdfhskjdhfk'); }).to.throw(Error, errorFormat);
      //expect(function(){tt.setCurrentDate('35 Dey 2981'); }).to.throw(Error, errorFormat);
      //expect(tt.setDate(new Date(), '28 Nov 2016').dateFormat).to.deep.equal('28 Nov 2016');
      expect(tt.setDate('28 Nov 2016').date).to.deep.equal('28 Nov 2016');
    });

    it('should allow to set its scope in hours with maximum 24 hours span', function () {
      var errorMsg = 'Timetable scope should consist of (start, end) in whole hours from 0 to 23';
      var tt = this.timetable;
      expect(tt.setScope(1, 20).scope).to.deep.equal({hourStart: 1, hourEnd: 20});
      expect(tt.setScope(0, 0).scope).to.deep.equal({hourStart: 0, hourEnd: 0});
      expect(function(){tt.setScope(1, 24);}).to.throw(RangeError, errorMsg);
      expect(function(){tt.setScope(-1, 1);}).to.throw(RangeError, errorMsg);
      expect(function(){tt.setScope(1, 0);}).not.to.throw(RangeError, errorMsg);
      expect(function(){tt.setScope(3, 1);}).not.to.throw(RangeError, errorMsg);
      expect(function(){tt.setScope(20);}).to.throw(RangeError, errorMsg);
      expect(function(){tt.setScope(undefined);}).to.throw(RangeError, errorMsg);
    });

    it('should allow to add a list of locations', function () {
      var tt = this.timetable;

      expect(tt.addLocations(['loc1', 'loc2']).locations).to.deep.equal(['loc1', 'loc2']);
      expect(tt.addLocations(['loc3']).locations).to.deep.equal(['loc1', 'loc2', 'loc3']);
      expect(function(){tt.addLocations(['loc3']);}).to.throw('Location already exists');
      expect(function(){tt.addLocations(undefined);}).to.throw('Tried to add locations in wrong format');
      expect(function(){tt.addLocations({});}).to.throw('Tried to add locations in wrong format');
    });

    it('should be possible to add events for an existing location using js dates', function () {
      var tt = this.timetable;
      tt.addLocations(['loc1', 'loc2', 'loc5']);

      var optionsWithURL = {
        url: '#'
      };

      var optionsWithClass = {
        class: 'vip-only'
      };

      var optionsWithURLAndClass = {
        url: '#',
        class: 'vip-only'
      };

      var optionsWithDataAttributes = {
        url: '#',
        data: {
          maxPlayers: '12',
          gameType: 'Muskets'
        }
      };

      expect(tt.addEvent('Zumba', 'loc2', new Date(2015,7,17,9), new Date(2015,7,17,17), optionsWithURL).events[0]).to.deep.equal({name: 'Zumba', location: 'loc2', startDate: new Date(2015,7,17,9), endDate: new Date(2015,7,17,17), options: { url: '#' }});
      expect(tt.addEvent('Private concert', 'loc1', new Date(2015,7,17,9), new Date(2015,7,17,17), optionsWithClass).events[1]).to.deep.equal({name: 'Private concert', location: 'loc1', startDate: new Date(2015,7,17,9), endDate: new Date(2015,7,17,17), options: { class: 'vip-only' }});
      expect(tt.addEvent('DIY Fireworks', 'loc1', new Date(2015,7,17,10,30), new Date(2015,7,17,11,15), optionsWithURL).events[2]).to.deep.equal({name: 'DIY Fireworks', location: 'loc1', startDate: new Date(2015,7,17,10,30), endDate: new Date(2015,7,17,11,15), options: { url: '#' }});
      expect(tt.addEvent('Afterparty', 'loc2', new Date(2015,7,17,22,30), new Date(2015,7,17,23,15), optionsWithURLAndClass).events[3]).to.deep.equal({name: 'Afterparty', location: 'loc2', startDate: new Date(2015,7,17,22,30), endDate: new Date(2015,7,17,23,15), options: { url: '#', class: 'vip-only' }});
      expect(tt.addEvent('DIY Fireworks', 'loc5', new Date(2015,7,17,20,30), new Date(2015,7,18,0,15)).events[4]).to.deep.equal({name: 'DIY Fireworks', location: 'loc5', startDate: new Date(2015,7,17,20,30), endDate: new Date(2015,7,18,0,15), options: undefined});
      expect(tt.addEvent('City Tour', 'loc1', new Date(2015,7,17,20,30), new Date(2015,7,18,0,15), 'some string').events[5]).to.deep.equal({name: 'City Tour', location: 'loc1', startDate: new Date(2015,7,17,20,30), endDate: new Date(2015,7,18,0,15), options: undefined});
      expect(tt.addEvent('Pub Crawl', 'loc1', new Date(2015,7,17,21,30), new Date(2015,7,18,2,0), []).events[6]).to.deep.equal({name: 'Pub Crawl', location: 'loc1', startDate: new Date(2015,7,17,21,30), endDate: new Date(2015,7,18,2,0), options: undefined });
      expect(tt.addEvent('Pub Crawl', 'loc1', new Date(2015,7,17,21,30), new Date(2015,7,18,2,0), optionsWithDataAttributes).events[7]).to.deep.equal({name: 'Pub Crawl', location: 'loc1', startDate: new Date(2015,7,17,21,30), endDate: new Date(2015,7,18,2,0), options: { url: '#', data: { maxPlayers: '12', gameType: 'Muskets' } } });
      expect(function(){tt.addEvent('DIY Fireworks', 'Area 51', new Date(2015,7,17,10), new Date(2015,7,17,11), optionsWithURL);}).to.throw('Unknown location');
      expect(function(){tt.addEvent('DIY Fireworks', 'loc2', new Date(2015,7,17,10,1), new Date(2015,7,17,10), optionsWithURL);}).to.throw('Invalid time range');
      expect(function(){tt.addEvent('DIY Fireworks', 'loc2', new Date(2015,7,17,10), new Date(2015,7,17,9), optionsWithURL);}).to.throw('Invalid time range');

    });

  });

})();


test/spec/plugin.js
(function() {
    'use strict';

    beforeEach(function() {
        this.timetable = new Timetable();
    });

    describe('Timetable', function() {

        it('should by default set its scope from 9 to 17 hours', function() {
            expect(this.timetable.scope).to.deep.equal({ hourStart: 9, hourEnd: 17 });
        });


        it('should by default set its time interval to 60 minutes', function() {
            expect(this.timetable.interval).equal(60);
        });

        it('should allow to set its interval from 1 to 60', function() {
            var errorMsg = 'Timetable interval should be an integer 1 to 60';
            var tt = this.timetable;
            expect(tt.setInterval(1).interval).equal(1);
            expect(tt.setInterval(60).interval).equal(60);
            expect(function() { tt.setInterval(0); }).to.throw(Error, errorMsg);
            expect(function() { tt.setInterval(61); }).to.throw(Error, errorMsg);
        });

        it('should allow to set its scope in hours with maximum 24 hours span', function() {
            var errorMsg = 'Timetable scope should consist of (start, end) in whole hours from 0 to 23';
            var tt = this.timetable;
            expect(tt.setScope(1, 20).scope).to.deep.equal({ hourStart: 1, hourEnd: 20 });
            expect(tt.setScope(0, 0).scope).to.deep.equal({ hourStart: 0, hourEnd: 0 });
            expect(function() { tt.setScope(1, 24); }).to.throw(RangeError, errorMsg);
            expect(function() { tt.setScope(-1, 1); }).to.throw(RangeError, errorMsg);
            expect(function() { tt.setScope(1, 0); }).not.to.throw(RangeError, errorMsg);
            expect(function() { tt.setScope(3, 1); }).not.to.throw(RangeError, errorMsg);
            expect(function() { tt.setScope(20); }).to.throw(RangeError, errorMsg);
            expect(function() { tt.setScope(undefined); }).to.throw(RangeError, errorMsg);
        });

        it('should allow to add a list of locations', function() {
            var tt = this.timetable;

            expect(tt.addLocations(['loc1', 'loc2']).locations).to.deep.equal(['loc1', 'loc2']);
            expect(tt.addLocations(['loc3']).locations).to.deep.equal(['loc1', 'loc2', 'loc3']);
            expect(function() { tt.addLocations(['loc3']); }).to.throw('Location already exists');
            expect(function() { tt.addLocations(undefined); }).to.throw('Tried to add locations in wrong format');
            expect(function() { tt.addLocations({}); }).to.throw('Tried to add locations in wrong format');
        });

        it('should be possible to add events for an existing location using js dates', function() {
            var tt = this.timetable;
            tt.addLocations(['loc1', 'loc2', 'loc5']);

            var optionsWithURL = {
                url: '#'
            };

            var optionsWithClass = {
                class: 'vip-only'
            };

            var optionsWithURLAndClass = {
                url: '#',
                class: 'vip-only'
            };

            var optionsWithDataAttributes = {
                url: '#',
                data: {
                    maxPlayers: '12',
                    gameType: 'Muskets'
                }
            };

            expect(tt.addEvent('Zumba', 'loc2', new Date(2015, 7, 17, 9), new Date(2015, 7, 17, 17), optionsWithURL).events[0]).to.deep.equal({ name: 'Zumba', location: 'loc2', startDate: new Date(2015, 7, 17, 9), endDate: new Date(2015, 7, 17, 17), options: { url: '#' } });
            expect(tt.addEvent('Private concert', 'loc1', new Date(2015, 7, 17, 9), new Date(2015, 7, 17, 17), optionsWithClass).events[1]).to.deep.equal({ name: 'Private concert', location: 'loc1', startDate: new Date(2015, 7, 17, 9), endDate: new Date(2015, 7, 17, 17), options: { class: 'vip-only' } });
            expect(tt.addEvent('DIY Fireworks', 'loc1', new Date(2015, 7, 17, 10, 30), new Date(2015, 7, 17, 11, 15), optionsWithURL).events[2]).to.deep.equal({ name: 'DIY Fireworks', location: 'loc1', startDate: new Date(2015, 7, 17, 10, 30), endDate: new Date(2015, 7, 17, 11, 15), options: { url: '#' } });
            expect(tt.addEvent('Afterparty', 'loc2', new Date(2015, 7, 17, 22, 30), new Date(2015, 7, 17, 23, 15), optionsWithURLAndClass).events[3]).to.deep.equal({ name: 'Afterparty', location: 'loc2', startDate: new Date(2015, 7, 17, 22, 30), endDate: new Date(2015, 7, 17, 23, 15), options: { url: '#', class: 'vip-only' } });
            expect(tt.addEvent('DIY Fireworks', 'loc5', new Date(2015, 7, 17, 20, 30), new Date(2015, 7, 18, 0, 15)).events[4]).to.deep.equal({ name: 'DIY Fireworks', location: 'loc5', startDate: new Date(2015, 7, 17, 20, 30), endDate: new Date(2015, 7, 18, 0, 15), options: undefined });
            expect(tt.addEvent('City Tour', 'loc1', new Date(2015, 7, 17, 20, 30), new Date(2015, 7, 18, 0, 15), 'some string').events[5]).to.deep.equal({ name: 'City Tour', location: 'loc1', startDate: new Date(2015, 7, 17, 20, 30), endDate: new Date(2015, 7, 18, 0, 15), options: undefined });
            expect(tt.addEvent('Pub Crawl', 'loc1', new Date(2015, 7, 17, 21, 30), new Date(2015, 7, 18, 2, 0), []).events[6]).to.deep.equal({ name: 'Pub Crawl', location: 'loc1', startDate: new Date(2015, 7, 17, 21, 30), endDate: new Date(2015, 7, 18, 2, 0), options: undefined });
            expect(tt.addEvent('Pub Crawl', 'loc1', new Date(2015, 7, 17, 21, 30), new Date(2015, 7, 18, 2, 0), optionsWithDataAttributes).events[7]).to.deep.equal({ name: 'Pub Crawl', location: 'loc1', startDate: new Date(2015, 7, 17, 21, 30), endDate: new Date(2015, 7, 18, 2, 0), options: { url: '#', data: { maxPlayers: '12', gameType: 'Muskets' } } });
            expect(function() { tt.addEvent('DIY Fireworks', 'Area 51', new Date(2015, 7, 17, 10), new Date(2015, 7, 17, 11), optionsWithURL); }).to.throw('Unknown location');
            expect(function() { tt.addEvent('DIY Fireworks', 'loc2', new Date(2015, 7, 17, 10, 1), new Date(2015, 7, 17, 10), optionsWithURL); }).to.throw('Invalid time range');
            expect(function() { tt.addEvent('DIY Fireworks', 'loc2', new Date(2015, 7, 17, 10), new Date(2015, 7, 17, 9), optionsWithURL); }).to.throw('Invalid time range');

        });

    });

})();

gulpfile.js
/*global -$ */
'use strict';
// generated on 2015-06-09 using generator-gulp-webapp 0.3.0
var gulp = require('gulp');
var $ = require('gulp-load-plugins')();
var browserSync = require('browser-sync');
var reload = browserSync.reload;

gulp.task('styles', function () {
  return gulp.src('app/styles/*.sass')
    .pipe($.sourcemaps.init())
    .pipe($.sass({
      outputStyle: 'nested', // libsass doesn't support expanded yet
      indentedSyntax: true,
      precision: 10,
      includePaths: ['.'],
      onError: console.error.bind(console, 'Sass error:')
    }))
    .pipe($.postcss([ require('autoprefixer')() ]))
    .pipe($.sourcemaps.write())
    .pipe(gulp.dest('.tmp/styles'))
    .pipe(reload({stream: true}));
});

gulp.task('jshint', function () {
  return gulp.src('app/scripts/**/*.js')
    .pipe(reload({stream: true, once: true}))
    .pipe($.jshint())
    .pipe($.jshint.reporter('jshint-stylish'))
    .pipe($.if(!browserSync.active, $.jshint.reporter('fail')));
});

gulp.task('html', ['styles'], function () {
  var assets = $.useref.assets({searchPath: ['.tmp', 'app', '.', 'bower_components']});

  return gulp.src('app/*.html')
    .pipe(assets)
    .pipe($.if('*.js', $.uglify()))
    .pipe($.if('*.css', $.postcss([ require('cssnano')() ])))
    .pipe(assets.restore())
    .pipe($.useref())
    .pipe(gulp.dest('dist'));
});

gulp.task('images', function () {
  return gulp.src('app/images/**/*')
    .pipe($.cache($.imagemin({
      progressive: true,
      interlaced: true,
      // don't remove IDs from SVGs, they are often used
      // as hooks for embedding and styling
      svgoPlugins: [{cleanupIDs: false}]
    })))
    .pipe(gulp.dest('dist/images'));
});

gulp.task('extras', function () {
  return gulp.src([
    'app/*.*',
    '!app/*.html'
  ], {
    dot: true
  }).pipe(gulp.dest('dist'));
});

gulp.task('clean', require('del').bind(null, ['.tmp', 'dist']));

gulp.task('serve', ['styles'], function () {
  browserSync({
    notify: false,
    port: 9000,
    server: {
      baseDir: ['.tmp', 'app', 'bower_components'],
      index: 'demo.html',
    }
  });

  // watch for changes
  gulp.watch([
    'app/*.html',
    'app/scripts/**/*.js',
    'app/images/**/*',
  ]).on('change', reload);

  gulp.watch('app/styles/**/*.sass', ['styles']);
});

gulp.task('build', ['jshint', 'html', 'images', 'extras'], function () {
  return gulp.src('dist/**/*').pipe($.size({title: 'build', gzip: true}));
});

gulp.task('default', ['clean'], function () {
  gulp.start('build');
});


public/scripts/timetable.js
/******/ (function(modules) { // webpackBootstrap
/******/ 	// The module cache
/******/ 	var installedModules = {};
/******/
/******/ 	// The require function
/******/ 	function __webpack_require__(moduleId) {
/******/
/******/ 		// Check if module is in cache
/******/ 		if(installedModules[moduleId]) {
/******/ 			return installedModules[moduleId].exports;
/******/ 		}
/******/ 		// Create a new module (and put it into the cache)
/******/ 		var module = installedModules[moduleId] = {
/******/ 			i: moduleId,
/******/ 			l: false,
/******/ 			exports: {}
/******/ 		};
/******/
/******/ 		// Execute the module function
/******/ 		modules[moduleId].call(module.exports, module, module.exports, __webpack_require__);
/******/
/******/ 		// Flag the module as loaded
/******/ 		module.l = true;
/******/
/******/ 		// Return the exports of the module
/******/ 		return module.exports;
/******/ 	}
/******/
/******/
/******/ 	// expose the modules object (__webpack_modules__)
/******/ 	__webpack_require__.m = modules;
/******/
/******/ 	// expose the module cache
/******/ 	__webpack_require__.c = installedModules;
/******/
/******/ 	// define getter function for harmony exports
/******/ 	__webpack_require__.d = function(exports, name, getter) {
/******/ 		if(!__webpack_require__.o(exports, name)) {
/******/ 			Object.defineProperty(exports, name, {
/******/ 				configurable: false,
/******/ 				enumerable: true,
/******/ 				get: getter
/******/ 			});
/******/ 		}
/******/ 	};
/******/
/******/ 	// getDefaultExport function for compatibility with non-harmony modules
/******/ 	__webpack_require__.n = function(module) {
/******/ 		var getter = module && module.__esModule ?
/******/ 			function getDefault() { return module['default']; } :
/******/ 			function getModuleExports() { return module; };
/******/ 		__webpack_require__.d(getter, 'a', getter);
/******/ 		return getter;
/******/ 	};
/******/
/******/ 	// Object.prototype.hasOwnProperty.call
/******/ 	__webpack_require__.o = function(object, property) { return Object.prototype.hasOwnProperty.call(object, property); };
/******/
/******/ 	// __webpack_public_path__
/******/ 	__webpack_require__.p = "";
/******/
/******/ 	// Load entry module and return exports
/******/ 	return __webpack_require__(__webpack_require__.s = 0);
/******/ })
/************************************************************************/
/******/ ([
/* 0 */
/***/ (function(module, exports, __webpack_require__) {

__webpack_require__(1);
module.exports = __webpack_require__(2);


/***/ }),
/* 1 */
/***/ (function(module, exports, __webpack_require__) {

"use strict";


window.Timetable = function () {
	this.scope = {
		hourStart: 9,
		hourEnd: 17
	};
	this.locations = [];
	this.events = [];
};

Timetable.Renderer = function (tt) {
	if (!(tt instanceof Timetable)) {
		throw new Error('Initialize renderer using a Timetable');
	}
	this.timetable = tt;
};

(function () {
	function isValidHourRange(start, end) {
		return isValidHour(start) && isValidHour(end);
	}
	function isValidHour(number) {
		return isInt(number) && isInHourRange(number);
	}
	function isInt(number) {
		return number === parseInt(number, 10);
	}
	function isInHourRange(number) {
		return number >= 0 && number < 24;
	}
	function locationExistsIn(loc, locs) {
		return locs.indexOf(loc) !== -1;
	}
	function isValidTimeRange(start, end) {
		var correctTypes = start instanceof Date && end instanceof Date;
		var correctOrder = start < end;
		return correctTypes && correctOrder;
	}
	function getDurationHours(startHour, endHour) {
		return endHour >= startHour ? endHour - startHour : 24 + endHour - startHour;
	}
	function isValidDate(dateString) {
		var isValid = false;
		var date;
		if (typeof dateString !== 'string') {
			return false;
		}
		date = new Date(dateString);

		if (Object.prototype.toString.call(date) === '[object Date]') {
			if (isNaN(date.getTime())) {
				return false;
			} else {
				//Need regExp check for month and day
				isValid = true;
			}
		} else {
			return false;
		}
		return isValid;
	}
	function prettyFormat(number) {
		var prefix = number < 10 ? '0' : '';
		return prefix + number;
	}
	//return the name of the month
	var monthes = 'Jan_Feb_Mar_Apr_May_Jun_Jul_Aug_Sep_Oct_Nov_Dec'.split('_');
	function localeMonths(month) {
		if (!month || month < 0 && month > 11) {
			throw new Error('Incorrect month.');
		}
		return monthes[month];
	}
	//Formate date to DD MMM YYYY
	function formatedDate(dateString) {
		if (!isValidDate(dateString)) {
			throw new Error('Incorrect date format.');
		}
		var date = new Date(dateString);
		return prettyFormat(date.getDate()) + ' ' + localeMonths(date.getMonth()) + ' ' + date.getFullYear();
	}

	Timetable.prototype = {
		options: {
			footer: false,
			header: false,
			date: '', //Use dateString in ISO8601 or IETF RFC 2822 format
			timeline: false,
			intevalId: ''
		},
		setScope: function (start, end) {
			if (isValidHourRange(start, end)) {
				this.scope.hourStart = start;
				this.scope.hourEnd = end;
			} else {
				throw new RangeError('Timetable scope should consist of (start, end) in whole hours from 0 to 23');
			}

			return this;
		},
		//Set current date to timetable
		setCurrentDate: function () {
			this.options.date = new Date().toDateString();
			return this.options.date;
		},
		//Set date to timetable
		setDate: function (date) {
			if (!isValidDate(date)) {
				throw new Error('Wrong date format. Use ISO-8601 or IETF RFC 2822 date format.');
			}
			this.options.date = date;
			return this.options.date;
		},
		//Set footer with time(clone header)
		setFooter: function () {
			this.options.footer = true;
		},
		setHeader: function () {
			this.options.header = true;
			if (this.options.date === '') {
				this.setCurrentDate();
			}
		},
		//Set timeline on
		setTimeline: function () {
			this.options.timeline = true;
		},
		getOptions: function () {
			return this.options;
		},
		addLocations: function (newLocations) {
			function hasProperFormat() {
				return newLocations instanceof Array;
			}

			var existingLocations = this.locations;

			if (hasProperFormat()) {
				newLocations.forEach(function (loc) {
					if (!locationExistsIn(loc, existingLocations)) {
						existingLocations.push(loc);
					} else {
						throw new Error('Location already exists');
					}
				});
			} else {
				throw new Error('Tried to add locations in wrong format');
			}

			return this;
		},
		addEvent: function (name, location, start, end, options) {
			if (!locationExistsIn(location, this.locations)) {
				throw new Error('Unknown location');
			}
			if (!isValidTimeRange(start, end)) {
				throw new Error('Invalid time range: ' + JSON.stringify([start, end]));
			}

			var optionsHasValidType = Object.prototype.toString.call(options) === '[object Object]';

			this.events.push({
				name: name,
				location: location,
				startDate: start,
				endDate: end,
				options: optionsHasValidType ? options : undefined,
				id: 'time-entry-' + id
			});

			id++;
			return this;
		},
		removeEvent: function (id) {
			var index;
			for (var i = 0; i < this.events.length; i++) {
				if (this.events[i].id === id) {
					index = i;
				}
			}
			if (index > -1) {
				this.events.splice(index, 1);
			}
		}
	};

	function emptyNode(node) {
		while (node.firstChild) {
			node.removeChild(node.firstChild);
		}
	}

	function prettyFormatHour(hour) {
		var prefix = hour < 10 ? '0' : '';
		return prefix + hour + ':00';
	}

	var id = 0;

	Timetable.Renderer.prototype = {
		draw: function (selector) {
			function checkContainerPrecondition(container) {
				if (container === null) {
					throw new Error('Timetable container not found');
				}
			}
			function appendTimetableAside(container) {
				var asideNode = container.appendChild(document.createElement('aside'));
				var asideULNode = asideNode.appendChild(document.createElement('ul'));
				appendRowHeaders(asideULNode);
			}
			function appendRowHeaders(ulNode) {
				for (var k = 0; k < timetable.locations.length; k++) {
					var liNode = ulNode.appendChild(document.createElement('li'));
					var spanNode = liNode.appendChild(document.createElement('span'));
					spanNode.className = 'row-heading';
					spanNode.textContent = timetable.locations[k];
				}
			}
			function appendTimetableSection(container) {
				var sectionNode = container.appendChild(document.createElement('section'));
				var timeNode = sectionNode.appendChild(document.createElement('time'));
				appendColumnHeaders(timeNode);
				appendTimeRows(timeNode);
			}
			function appendColumnHeaders(node) {
				var headerNode = node.appendChild(document.createElement('header'));
				if (timetable.options.header) {
					appendColumnDateHeaders(headerNode);
				}
				var headerULNode = headerNode.appendChild(document.createElement('ul'));

				var completed = false;
				var looped = false;

				headerULNode.className = 'headerTimeUl';
				for (var hour = timetable.scope.hourStart; !completed;) {
					var liNode = headerULNode.appendChild(document.createElement('li'));
					var spanNode = liNode.appendChild(document.createElement('span'));
					spanNode.className = 'time-label';
					spanNode.textContent = prettyFormatHour(hour);

					if (hour === timetable.scope.hourEnd && (timetable.scope.hourStart !== timetable.scope.hourEnd || looped)) {
						completed = true;
					}
					if (++hour === 24) {
						hour = 0;
						looped = true;
					}
				}
			}
			function appendColumnDateHeaders(node) {
				var headerULNode = node.appendChild(document.createElement('ul'));
				node.parentNode.getElementsByTagName('header')[0].style.height = '92px';
				container.getElementsByTagName('aside')[0].style.marginTop = '92px';

				var liNodeDay = headerULNode.appendChild(document.createElement('li'));
				var divNodeDay = liNodeDay.appendChild(document.createElement('div'));
				var liNodeNextDay = headerULNode.appendChild(document.createElement('li'));
				var divNodeNextDay = liNodeNextDay.appendChild(document.createElement('div'));
				divNodeDay.className = 'time-label';
				divNodeDay.textContent = formatedDate(timetable.options.date);
				divNodeNextDay.className = 'time-label';
				headerULNode.className = 'headerDateUl';

				if (timetable.scope.hourEnd < timetable.scope.hourStart) {
					var nextDay = new Date(timetable.options.date);
					nextDay.setDate(nextDay.getDate() + 1);
					divNodeNextDay.textContent = formatedDate(nextDay.toString());
					liNodeDay.style.width = (24 - timetable.scope.hourStart) * 96 + 'px';
					liNodeNextDay.style.width = timetable.scope.hourEnd * 96 + 'px';
				} else {
					liNodeDay.style.width = getDurationHours * 96 + 'px';
				}
			}
			function appendTimeRows(node) {
				var ulNode = node.appendChild(document.createElement('ul'));
				var footer = timetable.options.footer ? node.appendChild(document.createElement('footer')) : false;
				if (footer) {
					footer.appendChild(node.getElementsByClassName('headerTimeUl')[0].cloneNode(true));
				}

				ulNode.className = 'room-timeline';
				for (var k = 0; k < timetable.locations.length; k++) {
					var liNode = ulNode.appendChild(document.createElement('li'));
					appendLocationEvents(timetable.locations[k], liNode); /**/
				}
				if (timetable.options.timeline) {
					timelineDraw(container);
				}
			}
			function appendLocationEvents(location, node) {
				for (var k = 0; k < timetable.events.length; k++) {
					var event = timetable.events[k];
					if (event.location === location) {
						appendEvent(event, node);
					}
				}
			}
			function appendEvent(event, node) {
				var hasOptions = event.options !== undefined;
				var hasURL,
				    hasAdditionalClass,
				    hasDataAttributes = false;

				if (hasOptions) {
					hasURL = event.options.url !== undefined ? true : false;
					hasAdditionalClass = event.options.class !== undefined ? true : false;
					hasDataAttributes = event.options.data !== undefined ? true : false;
				}

				var elementType = hasURL ? 'a' : 'span';
				var aNode = node.appendChild(document.createElement(elementType));
				var smallNode = aNode.appendChild(document.createElement('small'));
				aNode.title = event.name;

				if (hasURL) {
					aNode.href = event.options.url;
				}
				if (hasDataAttributes) {
					for (var key in event.options.data) {
						aNode.setAttribute('data-' + key, event.options.data[key]);
					}
				}

				aNode.className = hasAdditionalClass ? 'time-entry ' + event.options.class : 'time-entry';
				aNode.style.width = computeEventBlockWidth(event);
				aNode.style.left = computeEventBlockOffset(event);
				smallNode.textContent = event.name;
			}
			function timelineDraw(container) {
				var now = new Date();
				if (now.valueOf() - new Date(timetable.options.date).valueOf() > 24 * 60 * 60 * 1000) {
					console.log('The date is different from the current one.');
					return 'NCD';
				}
				var nowHour = now.getHours();
				var nowMinutes = now.getMinutes();
				var currDay = new Date(timetable.options.date).getDate();
				var startHour = timetable.scope.hourStart;
				var endHour = timetable.scope.hourStart < timetable.scope.hourEnd ? timetable.scope.hourEnd : timetable.scope.hourEnd + 24;
				if (nowHour <= startHour && nowHour >= 24 * (now.getDate() - currDay) + endHour) {
					console.log('End of time range');
					return 'EOF';
				}
				var ttHeight = window.getComputedStyle(container).height;
				var timeline = container.getElementsByTagName('time')[0].appendChild(document.createElement('div'));
				timeline.className = 'timeline';
				timeline.style.marginTop = '-' + ttHeight;
				timeline.style.height = ttHeight;
				timeline.style.left = (nowHour - startHour) * 96 + Math.round(96 / 60 * nowMinutes) + 'px';
				timetable.options.intervalID = setInterval(() => {
					if (parseFloat(timeline.style.left) < (endHour - startHour) * 96) {
						timeline.style.left = parseFloat(timeline.style.left) + Math.round(96 / 60) + 'px';
						console.log(timeline.style.left);
					} else {
						clearInterval(timetable.options.intervalID);
						return 'EOF';
					}
				}, 60000);
			}
			function computeEventBlockWidth(event) {
				var start = event.startDate;
				var end = event.endDate;
				var durationHours = computeDurationInHours(start, end);
				return durationHours / scopeDurationHours * 100 + '%';
			}
			function computeDurationInHours(start, end) {
				return (end.getTime() - start.getTime()) / 1000 / 60 / 60;
			}
			function computeEventBlockOffset(event) {
				var scopeStartHours = timetable.scope.hourStart;
				var eventStartHours = event.startDate.getHours() + event.startDate.getMinutes() / 60;
				var hoursBeforeEvent = getDurationHours(scopeStartHours, eventStartHours);
				return hoursBeforeEvent / scopeDurationHours * 100 + '%';
			}

			var timetable = this.timetable;
			var scopeDurationHours = getDurationHours(timetable.scope.hourStart, timetable.scope.hourEnd);
			var container = document.querySelector(selector);
			checkContainerPrecondition(container);
			emptyNode(container);
			appendTimetableAside(container);
			appendTimetableSection(container);
		}
	};
})();

module.exports.Timetable = Timetable;
module.exports.Timetable.Renderer = Timetable.Renderer;

/***/ }),
/* 2 */
/***/ (function(module, exports) {

// removed by extract-text-webpack-plugin

/***/ })
/******/ ]);